File: spring-webflux/webflux-fn/src/main/java/com/example/UserHandler.java
package com.example;

import static org.springframework.http.MediaType.APPLICATION_JSON;
import static org.springframework.web.reactive.function.server.RequestPredicates.accept;
import static org.springframework.web.reactive.function.server.RequestPredicates.contentType;
import static org.springframework.web.reactive.function.server.RequestPredicates.queryParam;
import static org.springframework.web.reactive.function.server.ServerResponse.badRequest;
import static org.springframework.web.reactive.function.server.ServerResponse.created;
import static org.springframework.web.reactive.function.server.ServerResponse.notFound;
import static org.springframework.web.reactive.function.server.ServerResponse.ok;

import java.net.URI;

import org.springframework.context.annotation.Bean;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.server.RouterFunction;
import org.springframework.web.reactive.function.server.RouterFunctions;
import org.springframework.web.reactive.function.server.ServerRequest;
import org.springframework.web.reactive.function.server.ServerResponse;

import lombok.Data;
import lombok.RequiredArgsConstructor;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Component
@RequiredArgsConstructor
public class UserHandler {
    private final UserRepository userRepository;

    @Bean
    public RouterFunction<ServerResponse> routes() {
        return RouterFunctions.route()
                              .path("/accounts", builder1 -> builder1
                                      .nest(accept(APPLICATION_JSON), builder2 -> builder2
                                              .GET("/{id}", this::findById)
                                              .GET(queryParam("name", param -> true), this::findByName)
                                              .GET(this::findAll))
                                      .POST(contentType(APPLICATION_JSON), this::create))
                              .build();
    }

    private Mono<ServerResponse> findAll(ServerRequest request) {
        return ok().body(userRepository.findAll(), User.class);
    }

    private Mono<ServerResponse> findById(ServerRequest request) {
        int id = Integer.parseInt(request.pathVariable("id"));
        return userRepository.findById(id)
                             .flatMap(user -> ok().bodyValue(user))
                             .switchIfEmpty(notFound().build());
    }

    private Mono<ServerResponse> findByName(ServerRequest request) {
        Flux<User> users = request.queryParam("name")
                                  .map(userRepository::findByName)
                                  .orElseGet(Flux::empty);
        return ok().body(users, User.class);
    }

    private Mono<ServerResponse> create(ServerRequest request) {
        Mono<User> user = request.bodyToMono(Request.class)
                                 .map(req -> new User(req.getName()));
        return userRepository.saveAll(user)
                             .single()
                             .flatMap(created -> created(URI.create("/accounts/" + created.getId())).build())
                             .switchIfEmpty(badRequest().build());
    }

    @Data
    private static class Request {
        private String name;
    }
}

File: spring-webflux/webflux-fn-kotlin/src/main/kotlin/com/example/UserHandler.kt
package com.example

import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.asFlow
import org.springframework.context.annotation.Bean
import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.server.*
import org.springframework.web.reactive.function.server.ServerResponse.ok

@Component
class UserHandler {
    @Bean
    fun routes() = coRouter {
        "/users".nest {
            GET("", ::findAll)
            GET("/{id}", ::findById)
        }
    }

    suspend fun findAll(request: ServerRequest): ServerResponse {
        val users: Flow<User> = (1..5).map { User(it, "name$it") }.asFlow()
        return ok().bodyAndAwait(users)
    }

    suspend fun findById(request: ServerRequest): ServerResponse {
        delay(1000)
        val id = request.pathVariable("id").toInt()
        return ok().bodyValueAndAwait(User(id, "name$id"))
    }

    data class User(val id: Int, val name: String)
}
